src/flag_gems/utils/pointwise_dynamic.py

```diff
@@ -761,64 +705,205 @@ class PointwiseDynamicFunction:
 
         assert isinstance(scalar_fn, JITFunction)
         self._scalar_fn = scalar_fn
         self._scalar_fn_cache_key = scalar_fn.cache_key
         self.pid = os.getpid()
-        self.lock = threading.Lock()
+
+        self.config: CodeGenConfig = config or CodeGenConfig(
+            512, (65536, 65536, 65536), 32, True, False
+        )
 
         # instantiated & cached overloads
-        self.overloads: Mapping[str, Callable] = {}
+        self.overloads: Mapping[int, Callable] = {}
 
     def __call__(self, *args, **kwargs):
-        # note: kwargs should not be used in JITFunction directly
-        key = f"{self.arg_key(*args)}"
-        cache = self.overloads
-        lock = self.lock
-
-        while key not in cache:
-            # generate file & import it
-            with lock:
-                if key in cache:
+        # inputs must be passed by position, outputs must be passed by keyword
+        ndim, args, kwargs = self.prepare_args(*args, **kwargs)
+        overload = self.instantiate(ndim)
+        out = overload(*args, **kwargs)
+        # NOTE: overload keeps the type of outputs:
+        # if a pre-defiend output is a Tensor or StridedBuffer, the corresponding
+        # output is also a Tensor StridedBuffer, respectively
+        # since prepare_args Wraps all the arguments, the outputs are all StridedBuffer
+        # but if manually instantiated overload is directly called, take care of
+        # that manually
+        return self._unwrap(out)
+
+    @staticmethod
+    def use_fast_path(tensors):
+        return all_the_same_shape(tensors) and (
+            all_c_contiguous(tensors)
+            or (
+                all_the_same_stride(tensors)
+                and torch.ops.aten.is_non_overlapping_and_dense(tensors[0])
+            )
+        )
+
+    def prepare_args(self, *args, **kwargs):
+        # output allocation(when needed)
+        # task simplification & task-rank infernece & input-output reinterpretation
+        schema = self.fx
+        outputs_that_need_allocation: List[int] = []
+        out_tensors = []
+        for i in range(schema.num_output_tensors()):
+            k = f"out{i}"
+            if k in kwargs:
+                out_tensors.append(kwargs[k])
+            else:
+                outputs_that_need_allocation.append(i)
+        # input arguments must be passed by position
+        in_tensors = [item for i, item in enumerate(args) if schema.is_tensor(i)]
+
+        # output dtype promotions
+        outputs_dtypes_for_allocation = []
+        for i in outputs_that_need_allocation:
+            *arg_indices, method = schema._promotion_methods[i]
+            promote_args = (args[j] for j in arg_indices)
+            _, dtype = type_promotion(*promote_args, type_promotion=method)
+            outputs_dtypes_for_allocation.append(dtype)
+
+        tensors = out_tensors + in_tensors
+        if self.use_fast_path(tensors):  # dimension collapse & use physical ordering
```

Comment:
I have once implemented this `partial dimension collapsing` mechanism but removed it later since it adds more complexity to output layout inference & task simplification(mainly due to development complexity, since I tried to generate the code for output layout inference & task simplification instead of adding a function that handles the general cases for pointwise operation. As a result, runtime overhead is not yet profiled since I finally removed this feature). 

Task simplification involves dimension reordering & collapsing, which requires some analyses of shape & strides of each tensor operand, mainly sorting.

We can try adding it back & profile to see whether it is beneficial.

Reply:
It may not have a significant impact on performance, but if decide to enable tuning, it could affect the size of the tuning space due to the addition of an extra dimension.

Reply:
I see. Now we have a fast_that that is used when (all operands have the same shape & all non overlapping & the same stride order). This condition can be checked with relatively low cost.

We will add dimension reordering and collapsing back, since it would not add overhead if the fast path is used.